packages/flutter/lib/src/gestures/force_press.dart

```diff
@@ -0,0 +1,260 @@
+  }) : assert(startPressure != null),
+       assert(peakPressure != null),
+       super(debugOwner: debugOwner);
+
+  /// A pointer is in contact with the screen and has just pressed with a force
+  /// exceeding the [startPressure].
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressStartCallback onStart;
+
+  /// A pointer is in contact with the screen and is either moving on the plane
+  /// of the screen, pressing the screen with varying forces or both
+  /// simultaneously.
+  ///
+  /// This callback will be invoked for every frame after the invocation of
+  /// [onStart] no matter what the pressure is during this time period. The
+  /// position and pressure of the pointer is provided in the callback's
+  /// `details` argument, which is a [ForcePressUpdateDetails] object.
+  GestureForcePressUpdateCallback onUpdate;
+
+  /// A pointer is in contact with the screen and has just pressed with a force
+  /// exceeding the [peakPressure].
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressPeakCallback onPeak;
+
+  /// A pointer is no longer in contact with the screen.
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressEndCallback onEnd;
+
+  /// The pressure of the press required to initiate a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure.
+  double startPressure;
+
+  /// The pressure of the press required to peak a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure.
+  double peakPressure;
+
+  Duration _lastTimeStamp;
+  Offset _lastPosition;
+  _ForceState _state = _ForceState.ready;
+
+  @override
+  void addPointer(PointerEvent event) {
+    startTrackingPointer(event.pointer);
+    if (_state == _ForceState.ready) {
+      _state = _ForceState.possible;
+      _lastPosition = event.position;
+    }
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    assert(_state != _ForceState.ready);
+    // A static pointer with changes in pressure creates PointerMoveEvent events.
+    if (event is PointerMoveEvent || event is PointerDownEvent) {
+      final double pressure = _inverseLerp(event.pressureMin, event.pressureMax, event.pressure);
+      if (_state == _ForceState.possible) {
+        _lastTimeStamp = event.timeStamp;
+        _lastPosition = event.position;
+        if (pressure > startPressure) {
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
+      } else if (_state == _ForceState.accepted || _state == _ForceState.peaked || _state == _ForceState.started) {
```

Comment:
Having 2 layers of long conditionals here and in the next line is hard to read. 

Maybe just deal with if possible, then return since I don't think it's possible to do more in the same call.

Then add a few sequences of ifs so they're not nested.

Reply:
So the only reason not to return is if we want to call onUpdate the same frame as onStart. Which seems to be a good move? Would it be ok to just separate them into two different ifs but not return?